Support `tcat` notation in spikeGLX filename parsing

Hi,
I am currently using GLX extractor, for posterior kilosort3 processing. The spikeinterface version is 0.98.1.
The read_spikeglx function throws the following error.

```
In [6]: ls /mnt/boninlab/boninlabwip2024/data/ephys/shahriar/Raw/Chronic/SH050/230911/trof/
sh0123br200bB_g0_tcat.imec0.ap.bin*   sh0123br200_bigopenfield_dim_g0_tcat.imec0.ap.bin*
sh0123br200bB_g0_tcat.imec0.ap.meta*  sh0123br200_bigopenfield_dim_g0_tcat.imec0.ap.meta*

 In [7]: spikeinterface.__version__
Out[7]: '0.98.1'

In [8]: recordings = se.read_spikeglx( Path("/mnt/boninlab/boninlabwip2024/data/ephys/shahriar/Raw/Chronic/SH050/230911/trof/") , stream_id = "imec0.ap")

---------------------------------------------------------------------------

AssertionError                            Traceback (most recent call last)

Cell In[8], line 1

----> 1 recordings = se.read_spikeglx( Path("/mnt/boninlab/boninlabwip2024/data/ephys/shahriar/Raw/Chronic/SH050/230911/trof/") , stream_id = "imec0.ap")

File /opt/anaconda/anaconda3/envs/spikeinterface_0.98.1/lib/python3.10/site-packages/spikeinterface/extractors/neoextractors/spikeglx.py:51, in SpikeGLXRecordingExtractor.__init__(self, folder_path, load_sync_channel, stream_id, stream_name, all_annotations)

49 def __init__(self, folder_path, load_sync_channel=False, stream_id=None, stream_name=None, all_annotations=False):

50     neo_kwargs = self.map_to_neo_kwargs(folder_path, load_sync_channel=load_sync_channel)
---> 51     NeoBaseRecordingExtractor.__init__(
     52         self, stream_id=stream_id, stream_name=stream_name, all_annotations=all_annotations, **neo_kwargs
     53     )
     55     # open the corresponding stream probe for LF and AP
     56     # if load_sync_channel=False
     57     if "nidq" not in self.stream_id and not load_sync_channel:
File /opt/anaconda/anaconda3/envs/spikeinterface_0.98.1/lib/python3.10/site-packages/spikeinterface/extractors/neoextractors/neobaseextractor.py:185, in NeoBaseRecordingExtractor.__init__(self, stream_id, stream_name, block_index, all_annotations, use_names_as_ids, **neo_kwargs)
    156 def __init__(
    157     self,
    158     stream_id: Optional[str] = None,
   (...)
    163     **neo_kwargs: Dict[str, Any],
    164 ) -> None:
    165     """
    166     Initialize a NeoBaseRecordingExtractor instance.
    167 
   (...)
    182 
    183     """
--> 185     _NeoBaseExtractor.__init__(self, block_index, **neo_kwargs)
    187     kwargs = dict(all_annotations=all_annotations)
    188     if block_index is not None:
File /opt/anaconda/anaconda3/envs/spikeinterface_0.98.1/lib/python3.10/site-packages/spikeinterface/extractors/neoextractors/neobaseextractor.py:25, in _NeoBaseExtractor.__init__(self, block_index, **neo_kwargs)
     23 def __init__(self, block_index, **neo_kwargs):
     24     if not hasattr(self, "neo_reader"):  # Avoid double initialization
---> 25         self.neo_reader = self.get_neo_io_reader(self.NeoRawIOClass, **neo_kwargs)
     27     if self.neo_reader.block_count() > 1 and block_index is None:
     28         raise Exception(
     29             "This dataset is multi-block. Spikeinterface can load one block at a time. "
     30             "Use 'block_index' to select the block to be loaded."
     31         )
File /opt/anaconda/anaconda3/envs/spikeinterface_0.98.1/lib/python3.10/site-packages/spikeinterface/extractors/neoextractors/neobaseextractor.py:64, in _NeoBaseExtractor.get_neo_io_reader(cls, raw_class, **neo_kwargs)
     62 neoIOclass = getattr(rawio_module, raw_class)
     63 neo_reader = neoIOclass(**neo_kwargs)
---> 64 neo_reader.parse_header()
     66 return neo_reader
File /opt/anaconda/anaconda3/envs/spikeinterface_0.98.1/lib/python3.10/site-packages/neo/rawio/baserawio.py:179, in BaseRawIO.parse_header(self)
    166 def parse_header(self):
    167     """
    168     This must parse the file header to get all stuff for fast use later on.
    169 
   (...)
    177 
    178     """
--> 179     self._parse_header()
    180     self._check_stream_signal_channel_characteristics()
File /opt/anaconda/anaconda3/envs/spikeinterface_0.98.1/lib/python3.10/site-packages/neo/rawio/spikeglxrawio.py:94, in SpikeGLXRawIO._parse_header(self)
     91 for info in self.signals_info_list:
     92     # key is (seg_index, stream_name)
     93     key = (info['seg_index'], info['stream_name'])
---> 94     assert key not in self.signals_info_dict
     95     self.signals_info_dict[key] = info
     97     # create memmap
AssertionError: 
In [9]: 
As shown in the ls output above, there are two data and two meta files in the input directory, which I aim to concatenate afterwards via: spikeinterface.concatenate_recordings
```

Could you please help me to understand what the problem is?
I’ve also attached the meta file.

Your assistance is greatly appreciated.

Thank you,
Shahriar.
[sh0123br200_bigopenfield_dim_g0_tcat.imec0.ap.meta.txt](https://github.com/NeuralEnsemble/python-neo/files/14125704/sh0123br200_bigopenfield_dim_g0_tcat.imec0.ap.meta.txt)
[sh0123br200bB_g0_tcat.imec0.ap.meta.txt](https://github.com/NeuralEnsemble/python-neo/files/14125705/sh0123br200bB_g0_tcat.imec0.ap.meta.txt)

--- a/neo/rawio/spikeglxrawio.py
+++ b/neo/rawio/spikeglxrawio.py
@@ -387,18 +387,20 @@ def parse_spikeglx_fname(fname):
     Consider the filenames: `Noise4Sam_g0_t0.nidq.bin` or `Noise4Sam_g0_t0.imec0.lf.bin`
     The filenames consist of 3 or 4 parts separated by `.`
     1. "Noise4Sam_g0_t0" will be the `name` variable. This choosen by the user at recording time.
-    2. "_g0_" is the "gate_num"
-    3. "_t0_" is the "trigger_num"
+    2. "g0" is the "gate_num"
+    3. "t0" is the "trigger_num"
     4. "nidq" or "imec0" will give the `device`
     5. "lf" or "ap" will be the `stream_kind`
-        `stream_name` variable is the concatenation of `device.stream_kind`
+       `stream_name` variable is the concatenation of `device.stream_kind`
+
+    If CatGT is used, then the trigger numbers in the file names ("t0"/"t1"/etc.)
+    will be renamed to "tcat". In this case, the parsed "trigger_num" will be set to "cat".
 
     This function is copied/modified from Graham Findlay.
 
     Notes:
-        * Sometimes the original file name is modified by the user and "_gt0_" or "_t0_"
+        * Sometimes the original file name is modified by the user and "_g0_" or "_t0_"
           are manually removed. In that case gate_name and trigger_num will be None.
-        * If tcat is used, then the trigger_num will be set to "cat".
 
     Parameters
     ---------
@@ -412,7 +414,7 @@ def parse_spikeglx_fname(fname):
     gate_num: int or None
         The gate identifier, e.g. 0.
     trigger_num: int | str or None
-        The trigger identifier, e.g. 1. If tcat is used, then the trigger_num will be set to "cat".
+        The trigger identifier, e.g. 1. If CatGT is used, then the trigger_num will be set to "cat".
     device: str
         The probe identifier, e.g. "imec2"
     stream_kind: str or None

--- a/neo/rawio/spikeglxrawio.py
+++ b/neo/rawio/spikeglxrawio.py
@@ -396,52 +396,60 @@ def parse_spikeglx_fname(fname):
     This function is copied/modified from Graham Findlay.
 
     Notes:
-       * Sometimes the original file name is modified by the user and "_gt0_" or "_t0_"
+        * Sometimes the original file name is modified by the user and "_gt0_" or "_t0_"
           are manually removed. In that case gate_name and trigger_num will be None.
+        * If tcat is used, then the trigger_num will be set to "cat".
 
     Parameters
     ---------
     fname: str
         The filename to parse without the extension, e.g. "my-run-name_g0_t1.imec2.lf"
+
     Returns
     -------
     run_name: str
         The run name, e.g. "my-run-name".
     gate_num: int or None
         The gate identifier, e.g. 0.
-    trigger_num: int or None
-        The trigger identifier, e.g. 1.
+    trigger_num: int | str or None
+        The trigger identifier, e.g. 1. If tcat is used, then the trigger_num will be set to "cat".
     device: str
         The probe identifier, e.g. "imec2"
     stream_kind: str or None
         The data type identifier, "lf" or "ap" or None
     """
-    r = re.findall(r"(\S*)_g(\d*)_t(\d*)\.(\S*).(ap|lf)", fname)
-    if len(r) == 1:
+    re_standard = re.findall(r"(\S*)_g(\d*)_t(\d*)\.(\S*).(ap|lf)", fname)
+    re_tcat = re.findall(r"(\S*)_g(\d*)_tcat.(\S*).(ap|lf)", fname)
+    re_nidq = re.findall(r"(\S*)_g(\d*)_t(\d*)\.(\S*)", fname)
+    if len(re_standard) == 1:
         # standard case with probe
-        run_name, gate_num, trigger_num, device, stream_kind = r[0]
+        run_name, gate_num, trigger_num, device, stream_kind = re_standard[0]
+    elif len(re_tcat) == 1:
+        # tcat case
+        run_name, gate_num, device, stream_kind = re_tcat[0]
+        trigger_num = "cat"
+    elif len(re_nidq) == 1:
+        # case for nidaq
+        run_name, gate_num, trigger_num, device = re_nidq[0]
+        stream_kind = None
     else:
-        r = re.findall(r"(\S*)_g(\d*)_t(\d*)\.(\S*)", fname)
-        if len(r) == 1:
-            # case for nidaq
-            run_name, gate_num, trigger_num, device = r[0]
-            stream_kind = None
+        # the naming do not correspond lets try something more easy
+        # example: sglx_xxx.imec0.ap
+        re_else = re.findall(r"(\S*)\.(\S*).(ap|lf)", fname)
+        re_else_nidq = re.findall(r"(\S*)\.(\S*)", fname)
+        if len(re_else) == 1:
+            run_name, device, stream_kind = re_else_nidq[0]
+            gate_num, trigger_num = None, None
+        elif len(re_else_nidq) == 1:
+            # easy case for nidaq, example: sglx_xxx.nidq
+            run_name, device = re_else_nidq[0]
+            gate_num, trigger_num, stream_kind = None, None, None
         else:
-            # the naming do not correspond lets try something more easy
-            # example: sglx_xxx.imec0.ap
-            r = re.findall(r"(\S*)\.(\S*).(ap|lf)", fname)
-            if len(r) == 1:
-                run_name, device, stream_kind = r[0]
-                gate_num, trigger_num = None, None
-            else:
-                # easy case for nidaq, example: sglx_xxx.nidq
-                r = re.findall(r"(\S*)\.(\S*)", fname)
-                run_name, device = r[0]
-                gate_num, trigger_num, stream_kind = None, None, None
+            raise ValueError(f"Cannot parse filename {fname}")
 
     if gate_num is not None:
         gate_num = int(gate_num)
-    if trigger_num is not None:
+    if trigger_num is not None and trigger_num != "cat":
         trigger_num = int(trigger_num)
 
     return (run_name, gate_num, trigger_num, device, stream_kind)

--- a/neo/test/rawiotest/test_spikeglxrawio.py
+++ b/neo/test/rawiotest/test_spikeglxrawio.py
@@ -38,7 +38,7 @@ class TestSpikeGLXRawIO(BaseTestRawIO, unittest.TestCase):
         "spikeglx/multi_trigger_multi_gate/CatGT/CatGT-C",
         "spikeglx/multi_trigger_multi_gate/CatGT/CatGT-D",
         "spikeglx/multi_trigger_multi_gate/CatGT/CatGT-E",
-        "spikeglx/multi_trigger_multi_gate/CatGT/SupercatA",
+        # "spikeglx/multi_trigger_multi_gate/CatGT/SupercatA",
     ]
 
     def test_with_location(self):

--- a/neo/test/rawiotest/test_spikeglxrawio.py
+++ b/neo/test/rawiotest/test_spikeglxrawio.py
@@ -32,6 +32,13 @@ class TestSpikeGLXRawIO(BaseTestRawIO, unittest.TestCase):
         "spikeglx/NP2_subset_with_sync",
         # NP-ultra
         "spikeglx/np_ultra_stub",
+        # CatGT
+        "spikeglx/multi_trigger_multi_gate/CatGT/CatGT-A",
+        "spikeglx/multi_trigger_multi_gate/CatGT/CatGT-B",
+        "spikeglx/multi_trigger_multi_gate/CatGT/CatGT-C",
+        "spikeglx/multi_trigger_multi_gate/CatGT/CatGT-D",
+        "spikeglx/multi_trigger_multi_gate/CatGT/CatGT-E",
+        "spikeglx/multi_trigger_multi_gate/CatGT/SupercatA",
     ]
 
     def test_with_location(self):

--- a/neo/test/rawiotest/test_spikeglxrawio.py
+++ b/neo/test/rawiotest/test_spikeglxrawio.py
@@ -33,12 +33,12 @@ class TestSpikeGLXRawIO(BaseTestRawIO, unittest.TestCase):
         # NP-ultra
         "spikeglx/np_ultra_stub",
         # CatGT
-        "spikeglx/multi_trigger_multi_gate/CatGT/CatGT-A",
-        "spikeglx/multi_trigger_multi_gate/CatGT/CatGT-B",
-        "spikeglx/multi_trigger_multi_gate/CatGT/CatGT-C",
-        "spikeglx/multi_trigger_multi_gate/CatGT/CatGT-D",
-        "spikeglx/multi_trigger_multi_gate/CatGT/CatGT-E",
-        # "spikeglx/multi_trigger_multi_gate/CatGT/SupercatA",
+        "spikeglx/multi_trigger_multi_gate/CatGT/CatGT-A/catgt_5-19-2022-CI0_g0",
+        "spikeglx/multi_trigger_multi_gate/CatGT/CatGT-B/catgt_5-19-2022-CI0_g0",
+        "spikeglx/multi_trigger_multi_gate/CatGT/CatGT-C/catgt_5-19-2022-CI0_g0",
+        "spikeglx/multi_trigger_multi_gate/CatGT/CatGT-D/catgt_5-19-2022-CI0_g0",
+        "spikeglx/multi_trigger_multi_gate/CatGT/CatGT-E/catgt_5-19-2022-CI0_g0",
+        "spikeglx/multi_trigger_multi_gate/CatGT/SupercatA/supercat_5-19-2022-CI0_g0",
     ]
 
     def test_with_location(self):

